### PR TITLE
Appsody init to support --stack-registry

### LINKED ISSUE #696: Appsody init should support --stack-registry flag
**Is your feature request related to a problem? Please describe.**
The `appsody deploy/build/run/test/debug` commands support a `--stack-registry` flag that allows overriding of the Stack registry URL.
The `appsody init` command does not support that flag - forcing a user that wants to override the registry URL to use the flag on every command.

**Describe the solution you'd like**
Appsody `init` should support the flag. When the flag is used, the CLI should store the registry URL in the `stack` element of the `.appsody-config.yaml` file.

### PR DESCRIPTION
**Pull request by chilanti**
Fixes #696

### CODE REVIEW

File: cmd/utils.go [hunk 1]
```diff
@@ -532,6 +537,42 @@ func saveProjectNameToConfig(projectName string, config *RootCommandConfig) erro
 	config.Info.log("Your Appsody project name has been set to ", projectName)
 	return nil
 }
+func setStackRegistry(stackRegistry string, config *RootCommandConfig) error {
+	valid, err := ValidateHostNameAndPort(stackRegistry)
```
Comment (kylegc) [inline cmd/utils.go]: OverrideStackRegistry also calls this, so it probably doesn't need to be here

File: cmd/utils.go [hunk 2]
```diff
@@ -532,6 +537,42 @@ func saveProjectNameToConfig(projectName string, config *RootCommandConfig) erro
...
+	if !valid {
+		return err
+	}
+
+	// update the in-memory project name
```
Comment (kylegc) [inline cmd/utils.go]: should this comment be removed?

File: cmd/init.go [hunk 3]
```diff
@@ -307,8 +307,15 @@ func install(config *initCommandConfig) error {
 	config.Info.log("Setting up the development environment")
 
 	// reset config.StackRegistry and get it again from the newly untarred .appsody-config.yaml
-	// this will need to be updated when we support --stack-registry on the init command
-	config.StackRegistry = ""
+
+	if config.StackRegistry != "" {
```
Comment (kylegc) [inline cmd/init.go]: I'm not convinced this is going to work. What if the stack template's `.appsody-config.yaml` already fully qualifies the stack registry and the user doesn't set `--stack-registry` flag? In this case we shouldn't modify the file, but I'm worried this code will change it.

File: cmd/init.go [hunk 4]
```diff
@@ -307,8 +307,15 @@ func install(config *initCommandConfig) error {
...
+		config.Debug.Log("The will be set to be: ", config.StackRegistry)
```
Comment (kylegc) [inline cmd/init.go]: A better debug message: "Setting the stack registry in .appsody-config.yaml to "

Comment (chilanti) [reply cmd/init.go]: Done.

File: cmd/init.go (hunk 3 above)
Comment (chilanti) [reply cmd/init.go]: I've done some significant refactoring - please take a look.

File: cmd/utils.go [hunk 5]
```diff
@@ -532,60 +517,115 @@ func saveProjectNameToConfig(projectName string, config *RootCommandConfig) erro
...
+		return "", errors.Errorf("Invalid stack image name detected in project config file: %s", stack)
+	}
+	return "", errors.New("No stack image name detected in project config file")
 
-		config.Debug.Logf("Project stack after override: %s is: %s", config.StackRegistry, projectConfig.Stack)
-		config.ProjectConfig = &projectConfig
+}
+func getProjectConfig(config *RootCommandConfig) (*ProjectConfig, error) {
+
+	projectConfig, err := getProjectConfigFileContents(config)
```
Comment (kylegc) [inline cmd/utils.go]: Lets put back the `config.ProjectConfig == nil` caching

**Pull request merged**